### PR TITLE
fixing DrtXpsApi failures

### PR DESCRIPTION
**Pull request by stevenbrix**
Fixes failures in DrtXpsApi as outlined in #1054

The implementation in CoreFx is more strict on ensuring that callers properly close streams before they are re-opened in update mode. this prevents multiple different streams from being updated at the same time since it uncompresses the data and stores it in memory

The callstacks we get is this:
```
>	System.IO.Compression.dll!System.IO.Compression.ZipArchiveEntry.OpenInUpdateMode() Line 700	C#
 	System.IO.Compression.dll!System.IO.Compression.ZipArchiveEntry.Open() Line 299	C#
 	System.IO.Packaging.dll!System.IO.Packaging.ZipStreamManager.Open(System.IO.Compression.ZipArchiveEntry zipArchiveEntry, System.IO.FileMode streamFileMode, System.IO.FileAccess streamFileAccess)	Unknown
 	System.IO.Packaging.dll!System.IO.Packaging.ZipPackagePart.GetStreamCore(System.IO.FileMode streamFileMode, System.IO.FileAccess streamFileAccess)	Unknown
 	System.IO.Packaging.dll!System.IO.Packaging.PackagePart.GetStream(System.IO.FileMode mode, System.IO.FileAccess access)	Unknown
 	ReachFramework.dll!System.Windows.Xps.Packaging.XpsDocument.IsSignable.get()	Unknown
 	ReachFramework.dll!System.Windows.Xps.Packaging.XpsDocument.SignDigitally(System.Security.Cryptography.X509Certificates.X509Certificate certificate, bool embedCertificate, System.Windows.Xps.Packaging.XpsDigSigPartAlteringRestrictions restrictions, string signatureId, bool testIsSignable)	Unknown
 	ReachFramework.dll!System.Windows.Xps.Packaging.XpsDocument.SignDigitally(System.Security.Cryptography.X509Certificates.X509Certificate certificate, bool embedCertificate, System.Windows.Xps.Packaging.XpsDigSigPartAlteringRestrictions restrictions)	Unknown
 	DRTXpsApi.dll!DRT.DigSignVerifyTestSuite.SignPackage() Line 1244	C#
 	DRTXpsApi.dll!DRT.DrtBase.RunNextTest() Line 2060	C#
 	DRTXpsApi.dll!DRT.DrtBase.RunNextTestOperation(object arg) Line 2004	C#
```

Here is the code that throws the exception in CoreFx:
https://github.com/dotnet/corefx/blob/master/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L697

### CODE REVIEW

File: src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs [hunk 1]
```diff
@@ -1,1278 +1,1282 @@
...
-            )
-        {
-            CommitInternal();
-        }
-
-        #endregion IDisposable implementation
-    }
-}
-
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (stevenbrix) [inline src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs]: weird, viewing the diff in VS looks normal, not sure why this looks like the whole doc changed? maybe line-endings?

Comment (rladuca) [reply src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs]: What's the actual change to look at?

**Comment by vatsan-madhavan** [comment]
Is there a corresponding change needed in [dotnet-wpf-test](https://dev.azure.com/dnceng/internal/_git/dotnet-wpf-test) ?

File: src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs [hunk 2]
```diff
@@ -1,1278 +1,1282 @@
...
+        Commit(
+            )
+        {
+            CommitInternal();
+        }
+
+        /// <summary>
+        /// This method closes streams and frees memory for this
+        /// fixed document.
+        ///
```
Comment (vatsan-madhavan) [inline src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs]: It's very hard to review this file - GitHub indicates that the whole file has changed. 
I suspect that the real change starts here.

**Comment by vatsan-madhavan** [comment]
lgtm

File: src/Microsoft.DotNet.Wpf/src/WindowsBase/MS/Internal/IO/Packaging/XmlDigitalSignatureProcessor.cs [hunk 3]
```diff
@@ -725,7 +725,10 @@ private SignedXml EnsureXmlSignatureParsed()
             IEnumerable<System.Security.Cryptography.Xml.Reference> objectReferences)
         {
             // don't overwrite
-            Debug.Assert(SignaturePart.GetStream().Length == 0, "Logic Error: Can't sign when signature already exists");
+            using (Stream stream = SignaturePart.GetStream())
```
Comment (rladuca) [inline src/Microsoft.DotNet.Wpf/src/WindowsBase/MS/Internal/IO/Packaging/XmlDigitalSignatureProcessor.cs]: This is a change in behavior.  The Debug.Assert would never evaluate this code before, now it evals even in release.

File: src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs (hunk 2 above)
Comment (rladuca) [inline src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs]: While we're swapping to usings, maybe we should add one for the stream and writer here?

File: src/Microsoft.DotNet.Wpf/src/WindowsBase/MS/Internal/IO/Packaging/XmlDigitalSignatureProcessor.cs (hunk 3 above)
Comment (rladuca) [reply src/Microsoft.DotNet.Wpf/src/WindowsBase/MS/Internal/IO/Packaging/XmlDigitalSignatureProcessor.cs]: You can pass Func<bool> here and put the entire using inside the assert.

Comment (vatsan-madhavan) [reply src/Microsoft.DotNet.Wpf/src/WindowsBase/MS/Internal/IO/Packaging/XmlDigitalSignatureProcessor.cs]: You can refactor this into a helper method like this: 

```C#
          [System.Diagnostics.Conditional("DEBUG")]
          void HeperMethod(PackagePart signaturePart)
           {
                   using (Stream stream = SignaturePart.GetStream())
                  {
                     Debug.Assert(stream.Length == 0, "Logic Error: Can't sign when signature already exists");
                   }
            }
```

File: src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs (hunk 2 above)
Comment (stevenbrix) [reply src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs]: yeah, that's what i commented on above. if you look at this PR in Visual Studio you can see the change properly. i'm not sure why it looks like this in GitHub

File: src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs (hunk 1 above)
Comment (stevenbrix) [reply src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs]: it's just wrapping a call to `PackagePart.GetStream` in a using statement

Here are some pics of the diff in VS that lets you see the changes properly:
image

image

**Comment by stevenbrix** [comment]
> Is there a corresponding change needed in [dotnet-wpf-test](https://dev.azure.com/dnceng/internal/_git/dotnet-wpf-test) ?

Yeah, see this commit here:
https://dev.azure.com/dnceng/internal/_git/dotnet-wpf-test/commit/cb5fee5a70d02034786d3629a1668e816b7995bc

File: src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs (hunk 2 above)
Comment (stevenbrix) [reply src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs]: good idea.

**Comment by stevenbrix** [comment]
> Seems good to me.
> 
> Can you just be sure the line endings in XpsFixedDocumentReaderWriter.cs are correct? If you turn off ignore whitespace in VS that lights up just like GitHub, so you have a line ending transformation there.

Ahh yeah ok i think i remember now. The line endings were LF, but when i went to stage the file with `git add` i got the following error:

```
PS D:\dev\src\dotnet\wpf\src\Microsoft.DotNet.Wpf\src\ReachFramework\Packaging> git add .\XpsFixedDocumentReaderWriter.cs
fatal: LF would be replaced by CRLF in src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedDocumentReaderWriter.cs
```

so i had to change the line endings to CRLF

**Pull request merged**